python/tvm/relax/frontend/legalizer.py

```diff
@@ -0,0 +1,136 @@
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import math
+
+import tvm
+from tvm import relax, te, topi
+from tvm.ir.module import IRModule
+from tvm.relax.block_builder import BlockBuilder
+
+
+@relax.expr_functor.mutator
+class OperatorLegalizer(relax.PyExprMutator):
+    def __init__(self, mod: IRModule) -> None:
+        super().__init__(mod)
+        self.mod_ = mod
+
+    def _convert_op(self, call: relax.Call) -> relax.Call:
+        bb: BlockBuilder = self.builder_
+        op = call.op
+        attrs = call.attrs
+        args = call.args
+
+        if op.name == "relax.nn.conv2d":
+            return bb.call_te(
+                topi.nn.conv2d,
+                args[0],
+                args[1],
+                attrs.strides,
+                attrs.padding,
+                attrs.dilation,
+                attrs.data_layout,
+                attrs.kernel_layout,
+            )
+        elif op.name == "relax.add":
+            return bb.call_te(topi.add, args[0], args[1])
+        elif op.name == "relax.subtract":
+            return bb.call_te(topi.subtract, args[0], args[1])
+        elif op.name == "relax.multiply":
+            return bb.call_te(topi.multiply, args[0], args[1])
+        elif op.name == "relax.divide":
+            return bb.call_te(topi.divide, args[0], args[1])
+        elif op.name == "relax.floor_divide":
+            return bb.call_te(topi.floor_divide, args[0], args[1])
+        elif op.name == "relax.sin":
+            return bb.call_te(topi.sin, args[0])
+        elif op.name == "relax.cos":
+            return bb.call_te(topi.cos, args[0])
+        elif op.name == "relax.sqrt":
+            return bb.call_te(topi.sqrt, args[0])
+        elif op.name == "relax.nn.relu":
+            return bb.call_te(topi.nn.relu, args[0])
+        elif op.name == "relax.nn.gelu":
+
+            def te_gelu(x):
+                return te.compute(
+                    x.shape,
+                    lambda *i: 0.5
+                    * x(*i)
+                    * (
+                        1
+                        + te.tanh(math.sqrt(2 / math.pi) * (x(*i) + 0.044715 * te.power(x(*i), 3)))
+                    ),
+                )
+
+            return bb.call_te(te_gelu, args[0])
+        elif op.name == "relax.nn.silu":
+            sig = bb.emit_te(topi.sigmoid, args[0])
+            return bb.call_te(topi.multiply, args[0], sig)
+        elif op.name == "relax.reshape":
+            return bb.emit_te(topi.reshape, args[0], args[1])
```

Comment:
I'm just curious why you can return a var here by calling emit_te. The return type should be Call

Reply:
Yeah I think they should all be `call_te`, except for the new intermediate results. Updated now.